packages/apollo-server-core/src/ApolloServer.ts

```diff
@@ -381,6 +383,32 @@ export class ApolloServerBase {
     // is populated accordingly.
     this.ensurePluginInstantiation(plugins);
 
+    // We handle signals if it was explicitly requested, or if we're not in a test
+    // and it wasn't explicitly turned off. (For backwards compatibility, we check
+    // both the top-level handleSignals and nested inside 'engine'.)
+    if (
+      handleSignals === true ||
+      (typeof this.config.engine === 'object' &&
+        this.config.engine.handleSignals === true) ||
+      (process.env.NODE_ENV !== 'test' &&
+        handleSignals !== false &&
+        (typeof this.config.engine !== 'object' ||
+          this.config.engine.handleSignals !== false))
+    ) {
+      const signals: NodeJS.Signals[] = ['SIGINT', 'SIGTERM'];
+      signals.forEach((signal) => {
+        // Note: Node only started sending signal names to signal events with
+        // Node v10 so we can't use that feature here.
+        const handler: NodeJS.SignalsListener = async () => {
+          await this.stop();
+          process.kill(process.pid, signal);
```

Comment:
The theory is that we are not trying to prevent the process from dying when asked to, but just let it do some work before dying as requested.

It's not really a primitive that composes well, but I don't know of a better alternative.